judge/views/api/api_v1.py

```diff
@@ -109,18 +109,34 @@ def api_v1_user_list(request):
         'rank': rank
     } for username, name, points, rank in queryset})
 
-
 def api_v1_user_info(request, user):
     profile = get_object_or_404(Profile, user__username=user)
     submissions = list(Submission.objects.filter(case_points=F('case_total'), user=profile, problem__is_public=True, problem__is_organization_private=False)
                        .values('problem').distinct().values_list('problem__code', flat=True))
-    return JsonResponse({
+    resp = {
         'display_name': profile.name,
         'points': profile.points,
         'rank': profile.display_rank,
         'solved_problems': submissions,
         'organizations': list(profile.organizations.values_list('key', flat=True)),
-    })
+    }
+
+    last_rating = list(profile.ratings.order_by('-contest__end_time'))
+
+    contest_history = {}
+    for participation in (ContestParticipation.objects.filter(user=profile, virtual=0, contest__is_public=True, contest__is_private=False)
+                                  .order_by('-contest__end_time')):
+        contest_history[participation.contest.key] = {
+            'rating': participation.rating.rating if hasattr(participation, 'rating') else None,
```

Comment:
This and the line above will execute a query for every contest the user has participated in (= potentially many queries). You probably want to use something like `values_list` here instead.